From 3d7fa265f28676efcd9ba0f4620fd4a417989e9c Mon Sep 17 00:00:00 2001
From: Yusuke Goda <yusuke.goda.sx@renesas.com>
Date: Fri, 5 Jul 2013 18:09:23 +0900
Subject: [PATCH 616/675] ARM: shmobile: r8a7778: add USB Func support

This patch adds USB Function (= renesas_usbhs) driver support.
r8a7778_add_usb_phy_device()  should be called from .init_machine,
r8a7778_add_usb_func_device() should be called from .init_late
when platform use it.

This r8a7778_add_usb_func_device() will fill
renesas_usbhs_platform_info::platform_callback.get_id and
renesas_usbhs_platform_info::platform_callback.power_ctrl
if these callbacks were NULL, since these don't depend
on platform basically.

Signed-off-by: Kuninori Morimoto <kuninori.morimoto.gx@renesas.com>
---
 arch/arm/mach-shmobile/clock-r8a7778.c        |  1 +
 arch/arm/mach-shmobile/include/mach/r8a7778.h |  2 ++
 arch/arm/mach-shmobile/setup-r8a7778.c        | 40 ++++++++++++++++++++++++++-
 3 files changed, 42 insertions(+), 1 deletion(-)

diff --git a/arch/arm/mach-shmobile/clock-r8a7778.c b/arch/arm/mach-shmobile/clock-r8a7778.c
index 21fd068..122a55c 100644
--- a/arch/arm/mach-shmobile/clock-r8a7778.c
+++ b/arch/arm/mach-shmobile/clock-r8a7778.c
@@ -201,6 +201,7 @@ static struct clk_lookup lookups[] = {
 	CLKDEV_DEV_ID("rcar-du",	&mstp_clks[MSTP103]), /* DU */
 	CLKDEV_DEV_ID("ehci-platform", &mstp_clks[MSTP100]), /* USB EHCI port0/1 */
 	CLKDEV_DEV_ID("ohci-platform", &mstp_clks[MSTP100]), /* USB OHCI port0/1 */
+	CLKDEV_DEV_ID("renesas_usbhs", &mstp_clks[MSTP100]), /* USB FUNC */
 	CLKDEV_DEV_ID("i2c-rcar.0", &mstp_clks[MSTP030]), /* I2C0 */
 	CLKDEV_DEV_ID("i2c-rcar.1", &mstp_clks[MSTP029]), /* I2C1 */
 	CLKDEV_DEV_ID("i2c-rcar.2", &mstp_clks[MSTP028]), /* I2C2 */
diff --git a/arch/arm/mach-shmobile/include/mach/r8a7778.h b/arch/arm/mach-shmobile/include/mach/r8a7778.h
index be3f158..388081a 100644
--- a/arch/arm/mach-shmobile/include/mach/r8a7778.h
+++ b/arch/arm/mach-shmobile/include/mach/r8a7778.h
@@ -27,6 +27,7 @@
 #include <linux/platform_data/camera-rcar.h>
 #include <linux/platform_data/rcar-du.h>
 #include <linux/rcar-wdt.h>
+#include <linux/usb/renesas_usbhs.h>
 
 extern void r8a7778_add_standard_devices(void);
 extern void r8a7778_add_standard_devices_dt(void);
@@ -39,6 +40,7 @@ extern void r8a7778_add_rsnd_devices(struct rcar_snd_info *rsnd_info,
 				     int card_info_nr);
 extern void r8a7778_add_usb_phy_device(struct rcar_phy_platform_data *pdata);
 extern void r8a7778_add_usb_host_device(void);
+extern void r8a7778_add_usb_func_device(struct renesas_usbhs_platform_info *info);
 extern void r8a7778_add_vin_device(int id,
 				   struct rcar_vin_platform_data *pdata);
 extern void r8a7778_add_du_device(struct rcar_du_platform_data *info);
diff --git a/arch/arm/mach-shmobile/setup-r8a7778.c b/arch/arm/mach-shmobile/setup-r8a7778.c
index 66bf989..a7325d1 100644
--- a/arch/arm/mach-shmobile/setup-r8a7778.c
+++ b/arch/arm/mach-shmobile/setup-r8a7778.c
@@ -133,7 +133,7 @@ static int r8a7778_usb_phy_power(int enable)
 	return ret;
 }
 
-/* USB */
+/* USB HOST */
 static int usb_power_on(struct platform_device *pdev)
 {
 	int ret = r8a7778_usb_phy_power(1);
@@ -214,6 +214,44 @@ void __init r8a7778_add_usb_host_device(void)
 	platform_device_register_full(&ohci_info);
 }
 
+/* USB Func */
+static int usbhsf_get_id(struct platform_device *pdev)
+{
+	return USBHS_GADGET;
+}
+
+#define SUSPMODE	0x102
+static int usbhsf_power_ctrl(struct platform_device *pdev,
+			     void __iomem *base, int enable)
+{
+	enable = !!enable;
+
+	r8a7778_usb_phy_power(enable);
+
+	iowrite16(enable << 14, base + SUSPMODE);
+
+	return 0;
+}
+
+static struct resource usbhsf_resources[] __initdata = {
+	DEFINE_RES_MEM(0xffe60000, 0x110),
+	DEFINE_RES_IRQ(gic_iid(0x4f)),
+};
+
+void __init
+r8a7778_add_usb_func_device(struct renesas_usbhs_platform_info *info)
+{
+	if (!info->platform_callback.get_id)
+		info->platform_callback.get_id = usbhsf_get_id;
+	if (!info->platform_callback.power_ctrl)
+		info->platform_callback.power_ctrl = usbhsf_power_ctrl;
+
+	platform_device_register_resndata(&platform_bus, "renesas_usbhs", -1,
+					  usbhsf_resources,
+					  ARRAY_SIZE(usbhsf_resources),
+					  info, sizeof(*info));
+}
+
 /* Ether */
 static struct resource ether_resources[] __initdata = {
 	DEFINE_RES_MEM(0xfde00000, 0x400),
-- 
1.8.2

